File: csharp/Primes.cs
using System.Text;

public class Primes : Benchmark
{
    private const int PREFIX = 32338;
    private long _n;
    private long _prefix;
    private uint _result;

    public Primes()
    {
        _result = 5432;
        _n = ConfigVal("limit");
        _prefix = ConfigVal("prefix");
    }

    private sealed class Node
    {
        public readonly Node?[] Children = new Node?[10];
        public bool Terminal;

        public Node GetOrCreateChild(int digit)
        {
            ref Node? child = ref Children[digit];
            if (child is null) child = new Node();
            return child;
        }

        public Node? GetChild(int digit) => Children[digit];
    }

    private sealed class Sieve
    {
        private readonly int _limit;
        private readonly bool[] _isPrime;

        public Sieve(int limit)
        {
            _limit = limit;
            _isPrime = new bool[limit + 1];

            if (limit >= 2) Array.Fill(_isPrime, true, 2, _isPrime.Length - 2);
        }

        public List<int> CalculateAndGetPrimes()
        {
            if (_limit < 2) return new List<int>();

            int sqrtLimit = (int)Math.Sqrt(_limit);

            for (int p = 2; p <= sqrtLimit; p++)
            {
                if (_isPrime[p])
                {
                    for (int multiple = p * p; multiple <= _limit; multiple += p)
                    {
                        _isPrime[multiple] = false;
                    }
                }
            }

            int estimatedCount = _limit > 1000 ? (int)(_limit / Math.Log(_limit)) : _limit / 2;
            var primes = new List<int>(estimatedCount);

            if (_limit >= 2) primes.Add(2);

            for (int p = 3; p <= _limit; p += 2)
            {
                if (_isPrime[p]) primes.Add(p);
            }

            return primes;
        }
    }

    private static Node BuildTrie(List<int> primes)
    {
        var root = new Node();

        static void AddPrimeToTrie(Node root, int prime)
        {
            Node current = root;
            int temp = prime;
            Span<int> digits = stackalloc int[12];
            int digitCount = 0;

            while (temp > 0)
            {
                digits[digitCount++] = temp % 10;
                temp /= 10;
            }

            for (int i = digitCount - 1; i >= 0; i--)
            {
                current = current.GetOrCreateChild(digits[i]);
            }

            current.Terminal = true;
        }

        foreach (int prime in primes)
        {
            AddPrimeToTrie(root, prime);
        }

        return root;
    }

    private static List<int> FindWithPrefix(Node trieRoot, int prefix)
    {
        Node? current = trieRoot;
        int prefixValue = 0;

        int temp = prefix;
        Span<int> prefixDigits = stackalloc int[12];
        int digitCount = 0;

        while (temp > 0)
        {
            prefixDigits[digitCount++] = temp % 10;
            temp /= 10;
        }

        for (int i = digitCount - 1; i >= 0; i--)
        {
            int digit = prefixDigits[i];
            prefixValue = prefixValue * 10 + digit;

            current = current?.GetChild(digit);
            if (current is null) return new List<int>();
        }

        var results = new List<int>();
        var queue = new Queue<(Node, int)>();
        queue.Enqueue((current!, prefixValue));

        while (queue.Count > 0)
        {
            (Node node, int number) = queue.Dequeue();

            if (node.Terminal) results.Add(number);

            for (int digit = 0; digit < 10; digit++)
            {
                Node? child = node.GetChild(digit);
                if (child is not null) queue.Enqueue((child, number * 10 + digit));
            }
        }

        results.Sort();
        return results;
    }

    public override void Run(long IterationId)
    {
        var sieve = new Sieve((int)_n);
        var primes = sieve.CalculateAndGetPrimes();

        var trie = BuildTrie(primes);

        var results = FindWithPrefix(trie, (int)_prefix);

        _result += (uint)results.Count;
        foreach (int prime in results) _result += (uint)prime;
    }

    public override uint Checksum => _result;
}

File: fsharp/Primes.fs
namespace Benchmarks

open System
open System.Collections.Generic

module PrimeHelpers =
    [<AllowNullLiteral>]
    type Node() =
        let children = Array.zeroCreate<Node> 10
        let mutable terminal = false

        member _.Terminal 
            with get() = terminal 
            and set(v) = terminal <- v

        member this.GetOrCreateChild(digit: int) =
            match children.[digit] with
            | null ->
                let child = Node()
                children.[digit] <- child
                child
            | child -> child

        member _.GetChild(digit: int) = children.[digit]

    type Sieve = private { Limit: int; IsPrime: bool[] }

    module Sieve =
        let create limit =
            let isPrime = Array.create (limit + 1) true
            if limit >= 2 then
                Array.fill isPrime 2 (isPrime.Length - 2) true
            { Limit = limit; IsPrime = isPrime }

        let calculatePrimes sieve =
            if sieve.Limit < 2 then [] else

            let sqrtLimit = int (sqrt (float sieve.Limit))
            let isPrime = sieve.IsPrime

            seq { 2 .. sqrtLimit }
            |> Seq.filter (fun p -> isPrime.[p])
            |> Seq.iter (fun p ->
                seq { p * p .. p .. sieve.Limit }
                |> Seq.iter (fun multiple -> isPrime.[multiple] <- false))

            let primeCount = 
                if sieve.Limit > 1000 then 
                    int (float sieve.Limit / log (float sieve.Limit))
                else 
                    sieve.Limit / 2

            seq {
                yield 2
                for p in 3 .. 2 .. sieve.Limit do
                    if isPrime.[p] then yield p
            }
            |> Seq.toList

        let primes limit = 
            limit |> create |> calculatePrimes

    module Trie =
        let private getDigits num =
            let rec loop n acc =
                if n = 0 then acc
                else loop (n / 10) ((n % 10) :: acc)
            loop num []

        let build primes =
            let root = Node()

            let addPrime prime =
                let digits = getDigits prime
                digits
                |> List.fold (fun (node: Node) digit -> 
                    node.GetOrCreateChild(digit)) root
                |> fun node -> node.Terminal <- true

            primes |> List.iter addPrime
            root

        let findWithPrefix (root: Node) prefixValue =
            let getPrefixDigits prefix =
                let rec loop n acc =
                    if n = 0 then acc
                    else loop (n / 10) ((n % 10) :: acc)
                loop prefix []

            let prefixDigits = getPrefixDigits prefixValue

            let startNode = 
                prefixDigits
                |> List.fold (fun (node: Node) digit -> 
                    match node with
                    | null -> null
                    | n -> n.GetChild(digit)) root

            if isNull startNode then [] else

            let rec bfs (queue: Queue<Node * int>) (results: int list) =
                if queue.Count = 0 then results
                else
                    let (node, number) = queue.Dequeue()

                    let newResults = 
                        if node.Terminal then number :: results 
                        else results

                    let childrenResults =
                        [0..9]
                        |> List.fold (fun acc digit ->
                            let child = node.GetChild(digit)
                            if not (isNull child) then
                                queue.Enqueue(child, number * 10 + digit)
                            acc) newResults

                    bfs queue childrenResults

            let queue = Queue<Node * int>()
            queue.Enqueue(startNode, prefixValue)

            bfs queue []
            |> List.sort

type Primes() =
    inherit Benchmark()

    let mutable limit = 0L
    let mutable prefix = 0L
    let mutable checksum = 5432u

    let updateChecksum (results: int list) =
        checksum <- checksum + uint32 results.Length
        results 
        |> List.fold (fun sum prime -> sum + uint32 prime) checksum

    override this.Checksum = checksum

    override this.Prepare() =
        limit <- this.ConfigVal("limit")
        prefix <- this.ConfigVal("prefix")
        checksum <- 5432u

    override this.Run(_: int64) =
        let primes = PrimeHelpers.Sieve.primes (int limit)
        let trie = PrimeHelpers.Trie.build primes
        let results = PrimeHelpers.Trie.findWithPrefix trie (int prefix)
        checksum <- updateChecksum results